modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java

```diff
@@ -166,503 +152,196 @@ public int pages() {
-
-        try {
-            if (fileIo != this.fileIo) {
-                return;
-            }
-
-            try {
-                boolean interrupted = false;
-
-                while (true) {
-                    try {
-                        fileIo = null;
-
-                        fileIo = ioFactory.create(filePath, CREATE, READ, WRITE);
+    public void setCompleteCreationDeltaFilePageStoreIoCallback(CompleteCreationDeltaFilePageStoreIoCallback callback) {
+        completeCreationDeltaFilePageStoreIoCallback = callback;
+    }
 
-                        fileExists = true;
+    /**
+     * Gets or creates a new delta file, a new delta file will be created when the previous one is {@link #completeNewDeltaFile()
+     * completed}.
+     *
+     * <p>Thread safe.
+     *
+     * @param pageIndexesSupplier Page indexes supplier for the new delta file page store.
+     * @return Future that will be completed when the new delta file page store is created.
+     */
+    public CompletableFuture<DeltaFilePageStoreIo> getOrCreateNewDeltaFile(Supplier<int[]> pageIndexesSupplier) {
+        assert deltaFilePageStoreIoFactory != null;
 
-                        checkHeader(fileIo);
+        CompletableFuture<DeltaFilePageStoreIo> future = this.newDeltaFilePageStoreIoFuture;
 
-                        this.fileIo = fileIo;
+        if (future != null) {
+            return future;
+        }
 
-                        if (interrupted) {
-                            Thread.currentThread().interrupt();
-                        }
+        if (!NEW_DELTA_FILE_PAGE_STORE_IO_FUTURE.compareAndSet(this, null, future = new CompletableFuture<>())) {
+            // Another thread started creating a delta file.
+            return newDeltaFilePageStoreIoFuture;
+        }
 
-                        break;
-                    } catch (ClosedByInterruptException e) {
-                        interrupted = true;
+        int nextIndex = deltaFilePageStoreIos.isEmpty() ? 0 : deltaFilePageStoreIos.get(0).fileIndex() + 1;
 
-                        Thread.interrupted();
-                    }
-                }
-            } catch (IOException e) {
-                try {
-                    if (fileIo != null) {
-                        fileIo.close();
-                    }
-                } catch (IOException e0) {
-                    e.addSuppressed(e0);
-                }
+        DeltaFilePageStoreIo deltaFilePageStoreIo = deltaFilePageStoreIoFactory.create(nextIndex, pageIndexesSupplier.get());
 
-                throw e;
-            }
-        } finally {
-            readWriteLock.writeLock().unlock();
-        }
-    }
+        // Should add to the head, since read operations should always start from the most recent.
+        deltaFilePageStoreIos.add(0, deltaFilePageStoreIo);
 
-    private static int calcCrc32(ByteBuffer pageBuf, int pageSize) {
-        try {
-            pageBuf.position(0);
+        future.complete(deltaFilePageStoreIo);
```

Comment:
Future is completed here, is this correct?

Reply:
Yes